### ISSUE TITLE
[Proposal] introduction a BelongsToManyOverime functionnality

### ISSUE THREAD

**Issue by lk77** [laravel/internals#832]
Hello,

i have an idea,
it's kind of a new BelongsToManyOverTime relation, or an addition to

**Comment by 36864** [comment]
Essentially, this is just having a Many-to-Many relation where the timestamp fields are part of the pivot table's primary key?

I can see the use case on applications that keep an event log, but I'm not sure we need a new relation for this. I'm fairly sure this can be accomplished with a scope on your pivot model. Something along the lines of:
```
public function scopeActiveBetween($query, $date_start, $date_end)
{
  return $query->where('created_at', '<=', $date_start)
               ->where(function ($query) {
                  $query->orWhere('deleted_at', 'is', 'null')
                        ->orWhere('deleted_at' ,'>', $date_end)
                });
}
```

I guess a shortcut for this might come in handy in some cases, but I'm not sure it's something that needs to be in the core.

**Comment by lk77** [comment]
yes, you're right, for the created_at part it's certainly doable with scope, 
but i think it's become a bit messy if we add deleted_at part, 
because it will be full of unexpected behaviours. 

First i will need to use Model instead of Pivot, to be able to use SoftDelete
I will need to override SoftDeletingScope to handle onlyTrashed / withoutTrashed... because they will not work as expected, because deleted_at is not null even if it's not deleted yet (date in the future)

And i don't know how sync will handle this.

but yes, it's perhaps too specific to be added in the core

**Comment by michaeldyrynda** [comment]
I see the value of it, but I'd write it as a package that overrides the necessary functions and include it as a trait in the models that need it, rather than coming up with an entirely new model 👍
